lua/nvim-custom-diagnostic-highlight.lua

```diff
@@ -63,27 +101,67 @@ nvim_custom_diagnostic_highlight.setup = function(plugin_opts)
         )
 
         local diagnostic_namespace = vim.diagnostic.get_namespace(namespace)
-        if not diagnostic_namespace.user_data[final_opts.diagnostic_handler_namespace] then
-          diagnostic_namespace.user_data[final_opts.diagnostic_handler_namespace] = vim.api.nvim_create_namespace("")
-        end
+        local user_data = get_user_data(diagnostic_namespace, final_opts.diagnostic_handler_namespace)
 
         if should_highlight_diagnostic then
-          vim.highlight.range(
-            bufnr,
-            diagnostic_namespace.user_data[final_opts.diagnostic_handler_namespace],
-            higroup,
-            { diagnostic.lnum, diagnostic.col },
-            { diagnostic.end_lnum, diagnostic.end_col },
-            { priority = vim.highlight.priorities.diagnostics }
-          )
+          local set_highlight = function()
+            vim.highlight.range(
+              bufnr,
+              user_data.hl_namespace,
+              higroup,
+              { diagnostic.lnum, diagnostic.col },
+              { diagnostic.end_lnum, diagnostic.end_col },
+              { priority = vim.highlight.priorities.diagnostics }
+            )
+          end
+          local should_highlight = final_opts.defer_until_n_lines_away and
+            is_n_lines_away(bufnr, final_opts.defer_until_n_lines_away, diagnostic.lnum, diagnostic.end_lnum)
+
+          -- Defer if deferred highlighting is enabled and highlighting cannot be done now
+          -- Even here it's better to only check the current window, because user most likely
+          -- cares only for the position they are currently editing.
+          local defer = should_highlight and not should_highlight({ vim.api.nvim_get_current_win() })
+
+          if not defer then
+            set_highlight()
+          else
+            -- Store creates autocmds in user data to later delete those that didn't delete themselves
+            if not user_data.autocmds[bufnr] then
+              user_data.autocmds[bufnr] = {}
+            end
+            local autocmds_set = user_data.autocmds[bufnr]
+
+            local id
+            id = vim.api.nvim_create_autocmd('CursorMoved', {
```

Comment:
Maybe `CursorHold` makes more sense than `CursorMoved`, so the amount of calls is reduced a bit. Maybe another good approach would be to have the events listed in a table that comes from the options, something like `highlight_update_events = {'CursorHold', 'CursorHoldI'}`

Reply:
Sure, it's good idea to make the events configurable, I'll update the code.

Regarding computation overhead - it will probably be entirely unnoticeable, even on `CursorMoved,CursorMovedI`, because in essence it just runs [these 2 lines](https://github.com/Kasama/nvim-custom-diagnostic-highlight/pull/1/files#diff-17c39a1e5826e3aac48badbab83628c28e76f79d6461b77cc3f09a7ad7145d70R49-R50) per each deferred highlight (so in 90% - only once).